### DISCUSSION TITLE
Make HMC/MALA compatible with population MCMC

### DISCUSSION THREAD

**Discussion by ben18785** [pints-team/pints#1599] [Ideas/Science]
- At the moment, we don't have any methods to change the MCMC method within populationMCMC.
- It would be great if a user could specify `set_method(pints.HamiltonianMCMC)` here to use say HMC to sample from the various tempered distributions.
- This would require however that:
  - We create a `set_method` method for this class.
  - We have a method to modify if `self._needs_initial_phase:` to be changeable dependent on MCMC method we use
  - We would need to modify ask/tell to allow for a complete iteration of HMC (so that it actually returns a proposal rather than `None`).

The first two of these require changes to the PopulationMCMC class, the latter may be best handled by adding a method to HMC (I envision using complete iterations of HMC as intermediary distribution for quite a few cases). How about something like: `ask_complete()` and `tell_complete()`? I also like this idea as mostly users won't want the intermediate steps of HMC but the actual proposed values.

Thoughts?

**Comment by MichaelClerx** [comment]
Hi @ben18785 ! Thanks for the overview!

The third step isn't possible without breaking ask-and-tell. It would require the ask-complete method to do its own logpdf evaluations, which is explicitly the thing we're trying to avoid. Instead, the popmcmc class would simply have to return None as well every time the internal method returns None.
